Add suffixes for mimetypes to filepicker from python

toofar opened the pull request:
<!-- Thanks for submitting a pull request! Please pick a descriptive title (not just "issue 12345"). If there is an open issue associated to your PR, please add a line like "Closes #12345" somewhere in the PR description (outside of this comment) -->

Add filename extensions to the filepicker from the python `mimetypes` module to mitigate installs affected by the upstream bug around QMimeDatabase only looking at the first mime database where it found a registered type.

Oh, regarding the image/* bit, this is there qtwebengine does the same thing, for reference: https://github.com/qt/qtwebengine/blob/6.5.2/src/core/file_picker_controller.cpp#L264

Fixes: #7866

toofar commented:
Couple of questions around checking Qt versions, I might update the docstring for `qtutils.version_check()` based on the answers:

Regarding the `compiled` kwarg to `qtutils.version_check()`. In what situations would I want to use compiled=False? (I'm asking because I don't understand what it does.)

Also it seems the default behaviour is to check both the Qt version (two of them) and then the PyQt version. If I'm doing feature detection for a bug in Qt, I don't think I care about the PyQt version. So it seems like I could pass `compiled=False` to only check (one of) the Qt version(s). But from the name of the flag I'm not sure if that is the intended use case or not.

The-Compiler commented:
> Couple of questions around checking Qt versions, I might update the docstring for `qtutils.version_check()` based on the answers: [...]

The docstring could definitely do a better job at explaining this!

> Regarding the `compiled` kwarg to `qtutils.version_check()`. In what situations would I want to use compiled=False? (I'm asking because I don't understand what it does.)
> 
> Also it seems the default behaviour is to check both the Qt version (two of them) and then the PyQt version. If I'm doing feature detection for a bug in Qt, I don't think I care about the PyQt version. So it seems like I could pass `compiled=False` to only check (one of) the Qt version(s). But from the name of the flag I'm not sure if that is the intended use case or not.

It depends on what you want to check for:

- Want to use new API that appeared in, say, Qt 6.6? You need a PyQt 6.6 which is compiled against Qt 6.6 (rather than just having Qt 6.6 at runtime) -> `compiled=True`
- Want to check for a bug in Qt 6.6? You don't care about the PyQt version or what Qt it was compiled against, what happens is that you now have a Qt 6.6 at runtime -> `compiled=False`

That being said, I really don't like the API of `qtutils.version_check()` at all. I think it should be replaced by more explicit checks of some sort and a helper like `webengine_versions()`, which possibly has version numbers for `qt_runtime`, `qt_compiled`, `pyqt`, and `api` (which is basically `min(pyqt, qt_compiled)`).

The-Compiler commented on qutebrowser/browser/webengine/webview.py:
```diff
@@ -258,13 +259,51 @@ def acceptNavigationRequest(self,
+        Return any file extensions (aka suffixes) for mimetypes listed in
+        upstream_mimetypes that are not already contained in there.
+
+        # WORKAROUND: for https://bugreports.qt.io/browse/QTBUG-116905
+        # Affected Qt versions > 6.2.2 (probably) < 6.7.0
```
```suggestion
        WORKAROUND for https://bugreports.qt.io/browse/QTBUG-116905
        Affected Qt versions > 6.2.2 (probably) < 6.7.0
```

doesn't need a comment sign I guess?

The-Compiler commented on qutebrowser/browser/webengine/webview.py:
```diff
@@ -258,13 +259,51 @@ def acceptNavigationRequest(self,
+                accepted_mimetypes,
+                extra_suffixes,
+            )
+            accepted_mimetypes = list(accepted_mimetypes) + list(extra_suffixes)
```
I wonder if there is anything that depends on the order of those? From a quick peek at Qt's code I suppose it's fine if we don't keep the order, but I'm not sure.

toofar replied:
I can't think of anything, they are only used as an OR filter, not for any kind of first-match thing.

The-Compiler commented on qutebrowser/browser/webengine/webview.py:
```diff
@@ -258,13 +259,51 @@ def acceptNavigationRequest(self,
+        extra_suffixes = self.extra_suffixes_workaround(accepted_mimetypes)
+        if extra_suffixes:
+            log.webview.debug(
+                "adding extra suffixes to filepicker: before=%s added=%s",
```
As per https://github.com/qutebrowser/qutebrowser/pull/6197 we're mostly moving towards f-strings for logging, so that would probably make sense here too.

The-Compiler commented on qutebrowser/browser/webengine/webview.py:
```diff
@@ -258,13 +259,51 @@ def acceptNavigationRequest(self,
         self.navigation_request.emit(navigation)
         return navigation.accepted
 
+    @staticmethod
```
We don't really use staticmethods much. I wonder if this should just be a regular function in this module? No hard feelings either way though.

The-Compiler commented on tests/unit/browser/webengine/test_webview.py:
```diff
@@ -58,3 +61,78 @@ def test_enum_mappings(enum_type, naming, mapping):
+
+
+@pytest.mark.parametrize("before, extra", EXTRA_SUFFIXES_PARAMS)
+@mock.patch("qutebrowser.browser.webengine.webview.super")  # mock super() calls!
```
Huh what, that works!? Nice. Please use the [`mocker` fixture](https://pytest-mock.readthedocs.io/en/latest/) (as we already do elsewhere) though.

The-Compiler commented on tests/unit/browser/webengine/test_webview.py:
```diff
@@ -58,3 +61,78 @@ def test_enum_mappings(enum_type, naming, mapping):
+    before,
+    extra,
+):
+    # We can pass the class as "self" because we are only calling a static
```
Sneaky! If the workaround was a normal function, we could even pass `None` here, which I think would be slightly less confusing.

toofar commented:
Thanks for the review, I've addressed your comments and rebased to resolve a conflict with the changelog (and to remove the octothorps from the docstring).
